File: app/src/main/java/eu/michaelvogt/ar/author/utils/MarkerListAdapter.java
/*
    ARTester - AR for tourists by tourists
    Copyright (C) 2018  Michael Vogt

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package eu.michaelvogt.ar.author.utils;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import eu.michaelvogt.ar.author.R;
import eu.michaelvogt.ar.author.data.Marker;

public class MarkerListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
  private final static int TYPE_HEADER = 0;
  private final static int TYPE_ITEM = 1;

  private List<Marker> markers;
  private LayoutInflater inflater;

  private ItemClickListener listener;

  public void setItemClickListener(ItemClickListener listener) {
    this.listener = listener;
  }

  public MarkerListAdapter(Context context) {
    inflater = LayoutInflater.from(context);
  }

  public void setMarkers(List<Marker> markers) {
    this.markers = markers;
    notifyDataSetChanged();
  }

  @NonNull
  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    if (viewType == TYPE_HEADER) {
      View view = inflater.inflate(R.layout.card_header, parent, false);
      return new HeaderHolder(view);
    } else {
      View view = inflater.inflate(R.layout.card_marker, parent, false);
      return new MarkerHolder(view);
    }
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
    Marker item = markers.get(position);

    if (holder instanceof HeaderHolder) {
      HeaderHolder headerHolder = (HeaderHolder) holder;
      headerHolder.titleView.setText(item.getTitle());
    } else {
      MarkerHolder markerHolder = (MarkerHolder) holder;
      markerHolder.markerImageView.setImageBitmap(ImageUtils.decodeSampledBitmapFromImagePath(
          item.getMarkerImagePath(), 100, 100));
      markerHolder.nameView.setText(item.getTitle());
      markerHolder.infoView.setText(item.getIntro());
    }
  }

  @Override
  public int getItemViewType(int position) {
    return markers.get(position).isTitle() ? TYPE_HEADER : TYPE_ITEM;
  }

  @Override
  public int getItemCount() {
    return markers.size();
  }

  class MarkerHolder extends RecyclerView.ViewHolder {
    // each data item is just a string in this case
    private ImageView markerImageView;
    private TextView nameView;
    private TextView infoView;

    MarkerHolder(View view) {
      super(view);

      markerImageView = view.findViewById(R.id.location_image);
      nameView = view.findViewById(R.id.location_title);
      infoView = view.findViewById(R.id.marker_info);

      view.setOnClickListener(view1 -> {
        if (listener != null) {
          int position = getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            listener.onItemClicked(position);
          }
        }
      });
    }
  }

  class HeaderHolder extends RecyclerView.ViewHolder {
    private TextView titleView;

    HeaderHolder(View view) {
      super(view);

      titleView = view.findViewById(R.id.header_location);
    }
  }
}

File: app/src/main/java/eu/michaelvogt/ar/author/utils/MarkerListAdapter.kt
/*
    ARTester - AR for tourists by tourists
    Copyright (C) 2018  Michael Vogt

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package eu.michaelvogt.ar.author.utils

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import eu.michaelvogt.ar.author.R
import eu.michaelvogt.ar.author.data.Marker

class MarkerListAdapter(context: Context?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var markers: List<Marker> = ArrayList()
    private val inflater: LayoutInflater = LayoutInflater.from(context)

    private var listener: ItemClickListener? = null

    fun setItemClickListener(listener: ItemClickListener) {
        this.listener = listener
    }

    fun setMarkers(markers: List<Marker>) {
        this.markers = markers
        notifyDataSetChanged()
    }

    override
    fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == TYPE_HEADER) {
            val view = inflater.inflate(R.layout.card_header, parent, false)
            HeaderHolder(view)
        } else {
            val view = inflater.inflate(R.layout.card_marker, parent, false)
            MarkerHolder(view)
        }
    }

    override
    fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = markers[position]
        if (holder is HeaderHolder) {
            holder.titleView.text = item.title
        } else {
            val markerHolder = holder as MarkerHolder
            markerHolder.markerImageView.setImageBitmap(ImageUtils.decodeSampledBitmapFromImagePath(
                    item.markerImagePath, 100, 100))
            markerHolder.nameView.text = item.title
            markerHolder.infoView.text = item.intro
        }
    }

    override
    fun getItemViewType(position: Int): Int {
        return if (markers[position].isTitle()) TYPE_HEADER else TYPE_ITEM
    }

    override
    fun getItemCount(): Int {
        return markers.size
    }

    internal inner class MarkerHolder(view: View) : RecyclerView.ViewHolder(view) {
        internal val markerImageView: ImageView = view.findViewById(R.id.location_image)
        internal val nameView: TextView = view.findViewById(R.id.location_title)
        internal val infoView: TextView = view.findViewById(R.id.marker_info)

        init {
            view.setOnClickListener {
                if (listener != null) {
                    val position = adapterPosition
                    if (position != RecyclerView.NO_POSITION) {
                        val uId = markers[position].uId
                        listener!!.onItemClicked(uId)
                    }
                }
            }
        }
    }

    internal inner class HeaderHolder(view: View) : RecyclerView.ViewHolder(view) {
        internal val titleView: TextView = view.findViewById(R.id.header_location)
    }

    companion object {
        private const val TYPE_HEADER = 0
        private const val TYPE_ITEM = 1
    }
}